ciao-controller/client.go

```diff
@@ -114,26 +109,84 @@ func (client *ssntpClient) releaseResources(instanceID string) error {
 	return nil
 }
 
-func (client *ssntpClient) removeInstance(instanceID string) {
-	err := client.releaseResources(instanceID)
+func (client *ssntpClient) removeInstance(i *types.Instance) {
+	err := client.releaseResources(i)
 	if err != nil {
 		glog.Warningf("Error when releasing resources for deleted instance: %v", err)
 	}
-	client.deleteEphemeralStorage(instanceID)
-	err = client.ctl.ds.DeleteInstance(instanceID)
+	client.deleteEphemeralStorage(i.ID)
+	err = client.ctl.ds.DeleteInstance(i.ID)
 	if err != nil {
 		glog.Warningf("Error deleting instance from datastore: %v", err)
 	}
+
+	if i.CNCI {
+		return
+	}
+
+	// check the cnci after 5 minutes has elapsed to see if it's idle.
+	go func() {
+		time.Sleep(5 * time.Minute)
+		err = client.checkCNCI(i.TenantID)
+		if err != nil {
+			glog.Warningf("Error deleting CNCI: %v", err)
+		}
+	}()
+}
+
+func (client *ssntpClient) checkCNCI(tenantID string) error {
+	t, err := client.ctl.ds.GetTenant(tenantID)
+	if err != nil {
+		return errors.Wrapf(err, "error getting tenant from datastore")
+	}
+
+	cncis, err := client.ctl.ds.GetTenantCNCISummary(t.CNCIID)
+	if err != nil {
+		return errors.Wrapf(err, "error getting cnci from datastore")
+	}
+
+	// tenants only have one cnci right now
+	cnci := cncis[0]
+
+	if len(cnci.Subnets) > 0 {
```

Comment:
We did discuss using a go-routine to manage this stuff offline.  I was a bit hesitant about doing this when we spoke but now I better understand the problem, having seen the code.  One possible solution would be to have a  go routine per subnet that manages the CNCI lifecycle.   This would make it easy to

1. Protect against starting two CNCIs at one ( currently the job of confirmTenant ) 
2. Protect against removing a CNCI that has just been added.
3. Implement a timer for removing the CNCI when the last instance for that subnet has gone away.
4. Protect against deleting a CNCI that is starting up.  Maybe this isn't a problem.

We need to be careful to handle the controller restart case, i.e., it would need to start the go-routines for all the cncis on startup.